vhost-device-spi/src/spi.rs

```diff
@@ -0,0 +1,1119 @@
+        // SAFETY: Safe as the file is a valid SPI controller.
+        let ret = unsafe { ioctl_with_ref(&self.file.as_raw_fd(), SPI_IOC_WR_MODE32(), &mode) };
+
+        if ret == -1 {
+            Err(Error::IoctlFailure("set_mode", IoError::last()))
+        } else {
+            Ok(())
+        }
+    }
+
+    fn rdwr(&self, reqs: &mut [SpiTransReq]) -> Result<()> {
+        let mut msgs: Vec<SpiIocTransfer> = Vec::with_capacity(reqs.len());
+        let len = reqs.len();
+        let mut tx_buf_ptr: *mut u8;
+        let mut rx_buf_ptr: *mut u8;
+
+        let saved_mode: u32 = self.get_mode()?;
+        let mut trans_mode: u32 = saved_mode;
+
+        for req in reqs {
+            if req.tx_buf.is_empty() {
+                tx_buf_ptr = ptr::null_mut();
+            } else {
+                tx_buf_ptr = req.tx_buf.as_mut_ptr();
+            }
+
+            if req.rx_buf.is_empty() {
+                rx_buf_ptr = ptr::null_mut();
+            } else {
+                rx_buf_ptr = req.rx_buf.as_mut_ptr();
+            }
+
+            msgs.push(SpiIocTransfer {
+                tx_buf: tx_buf_ptr as u64,
+                rx_buf: rx_buf_ptr as u64,
+                len: req.trans_len,
+                speed_hz: req.speed_hz,
+                delay_usecs: req.delay_usecs,
+                bits_per_word: req.bits_per_word,
+                cs_change: req.cs_change,
+                tx_nbits: req.tx_nbits,
+                rx_nbits: req.rx_nbits,
+                word_delay_usecs: req.word_delay_usecs,
+                pad: 0,
+            });
+
+            if (req.mode & SPI_CPHA) == SPI_CPHA {
+                trans_mode |= LNX_SPI_CPHA;
+            } else {
+                trans_mode &= !LNX_SPI_CPHA;
+            }
+            if (req.mode & SPI_CPOL) == SPI_CPOL {
+                trans_mode |= LNX_SPI_CPOL;
+            } else {
+                trans_mode &= !LNX_SPI_CPOL;
+            }
+            if (req.mode & SPI_CS_HIGH) == SPI_CS_HIGH {
+                trans_mode |= LNX_SPI_CS_HIGH;
+            } else {
+                trans_mode &= !LNX_SPI_CS_HIGH;
+            }
+            if (req.mode & SPI_LSB_FIRST) == SPI_LSB_FIRST {
+                trans_mode |= LNX_SPI_LSB_FIRST;
+            } else {
+                trans_mode &= !LNX_SPI_LSB_FIRST;
+            }
+            if (req.mode & SPI_LOOP) == SPI_LOOP {
+                trans_mode |= LNX_SPI_LOOP;
+            } else {
+                trans_mode &= !LNX_SPI_LOOP;
+            }
+        }
+
+        self.set_mode(trans_mode)?;
+
+        // SAFETY: Safe as the file is a valid SPI controller.
+        let ret = unsafe {
+            ioctl_with_mut_ptr(
+                &self.file.as_raw_fd(),
+                spi_ioc_message(len as u32),
+                msgs.as_mut_ptr(),
+            )
+        };
+
+        self.set_mode(saved_mode)?;
+
+        if ret == -1 {
+            Err(Error::IoctlFailure("rdwr", IoError::last()))
```

Comment:
IIUC set_mode() calls another IOCTL, so IoError::last() will not contains the error we want to report here, should we move this block before calling set_mode()?

Maybe is better to wrap all the ioctl in safe functions and do there all the checks, otherwise it will become a mess.

Reply:
Yes! 
I will move the process in PhysDevice rdwr to SpiController transfer method, it will be much clearer.